sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -203,3 +203,28 @@ abstract class BinaryNode extends LogicalPlan {
 abstract class OrderPreservingUnaryNode extends UnaryNode {
   override final def outputOrdering: Seq[SortOrder] = child.outputOrdering
 }
+
+object LogicalPlanIntegrity {
+
+  private def canGetOutputAttrs(p: LogicalPlan): Boolean = {
+    p.resolved && !p.expressions.exists { e =>
+      // Some plans cannot call `output` because their expressions have `Unevaluable`,
+      // e.g., `Join` having a `ExistenceJoin` type.
+      e.collectFirst { case _: Unevaluable => true }.isDefined
+    }
+  }
+
+  /**
+   * This method checks if the same expression ID, `ExprId`, refer to an unique attribute.
+   * Some plan transformers (e.g., `RemoveNoopOperators`) rewrite logical
+   * plans based on this assumption.
+   */
+  def hasUniqueExprIdsForAttributes(plan: LogicalPlan): Boolean = {
+    val allOutputAttrs = plan.collect { case p if canGetOutputAttrs(p) =>
+      p.output.filter(_.resolved).map(_.canonicalized.asInstanceOf[Attribute])
+    }
+    val groupedAttrsByExprId = allOutputAttrs
+      .flatten.groupBy(_.exprId).values.map(_.distinct)
```

Comment:
what do we actually check? name, data type and nullability?

Reply:
Since references are canonicalized in advance, the current code just check a data type and a exprId:
https://github.com/apache/spark/blob/a1e459ed9f6777fb8d5a2d09fda666402f9230b9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala#L44-L45

Reply:
This reminds me of the hacks in `FoldablePropagation`.

We do have attributes with the same exprId but are actually different attributes, but I don't think there is an easy way to detect it automatically. For example, `a + 1 as a`, if I reuse the exprId of attribute `a` in the alias, how can we detect it? The name, data type and nullability are all the same.

Reply:
Oh, good point. Current approach only detects it once the data type is changed. But we generally reuse exprId on another attribute with same data type.